[ko] Translation for setInterval (#12091)
mdn/translated-content: files/ko/web/api/setinterval/index.md

## 예제

### 예제 1: 기본 구문

다음 예제는 `setInterval()`의 기본 구문을 보여줍니다.

```js
const intervalID = setInterval(myCallback, 500, 'Parameter 1', 'Parameter 2');

function myCallback(a, b)
{
 // 여기에 코드를 작성하세요
 // 매개변수는 순전히 선택적입니다.
 console.log(a);
 console.log(b);
}
```

### 예제 2: 두 가지 색상 변환하기

다음 에제는 중지 버튼을 누를 때까지 1초에 한 번씩 `flashtext()` 함수를 호출합니다.

#### HTML

```html
<div id="my_box">
  <h3>Hello World</h3>
</div>
<button id="start">Start</button>
<button id="stop">Stop</button>
```

#### CSS

```css
.go {
  color: green;
}
.stop {
  color: red;
}
```

#### JavaScript

```js
// intervalID를 저장할 변수입니다
let nIntervId;

function changeColor() {
  // 간격이 이미 설정되어 있는지 확인합니다
  if (!nIntervId) {
    nIntervId = setInterval(flashText, 1000);
  }
}

function flashText() {
  const oElem = document.getElementById("my_box");
  oElem.className = oElem.className === "go" ? "stop" : "go";
}

function stopTextColor() {
  clearInterval(nIntervId);
  // 변수에서 intervalID를 해체합니다
  nIntervId = null;
}

document.getElementById("start").addEventListener("click", changeColor);
document.getElementById("stop").addEventListener("click", stopTextColor);
```

#### 결과

{{EmbedLiveSample("Example_2:_Alternating_two_colors")}}

> 같이 보기: [`clearInterval()`](/en-US/docs/Web/API/clearInterval)

Reviewer: ```suggestion
같이 보기: [`clearInterval()`](/ko/docs/Web/API/clearInterval)
```
한글 문서가 존재하지 않더라도 /en-US 문서로 넘겨주기 때문에 한국 문서의 모든 경로는 /ko로 표기합니다. (참조 [번역 안내서](https://github.com/mdn/translated-content/blob/main/docs/ko/guides/translation-guide.md#%EB%AC%B8%EC%84%9C-%EC%97%B0%EA%B2%B0-%EA%B2%BD%EB%A1%9C) )

## "this" 문제

메서드를 `setInterval()` 또는 다른 함수에 전달하면 잘못된
[`this`](/ko/docs/Web/JavaScript/Reference/Operators/this) 값으로 호출됩니다.

> 이 문제는 [JavaScript reference](/en-US/docs/Web/JavaScript/Reference/Operators/this#as_an_object_method)에서 자세히 설명합니다.

Reviewer: ```suggestion
이 문제는 [JavaScript reference](/ko/docs/Web/JavaScript/Reference/Operators/this#as_an_object_method)에서 자세히 설명합니다.
```

### 설명

`setInterval()`에 의해 실행되는 코드는 호출된 함수와 별도의 실행 컨텍스트에서

> 실행됩니다. 결과적으로 호출된 함수의 [`this`](/en-US/docs/Web/JavaScript/Reference/Operators/this)

Reviewer: ```suggestion
실행됩니다. 결과적으로 호출된 함수의 [`this`](/ko/docs/Web/JavaScript/Reference/Operators/this)
```

키워드는 `window` (또는 `global`) 객체로 설정되며 `setTimeout`을 호출한 함수의 `this` 값과
동일하지 않습니다. `setTimeout()` 대신 `setInterval()` 을 사용하여
다음 예제를 참조하세요(사실 문제는 두 타이머 모두 동일합니다)

```js
myArray = ['zero', 'one', 'two'];

myArray.myMethod = function (sProperty) {
  alert(arguments.length > 0 ? this[sProperty] : this);
};

myArray.myMethod(); // "zero,one,two"가 출력됩니다
myArray.myMethod(1); // "one"가 출력됩니다
setTimeout(myArray.myMethod, 1000); // 1초 후 "[object Window]"가 출력됩니다
setTimeout(myArray.myMethod, 1500, "1"); // 1,5초 후에 "undefined"가 출력됩니다

// myArray.myMethod 내에서 this의 값을 변경하려는 동안
// setTimeout 내부에서 this의 값을 변경하기 때문에
// .call과 함께 'this'객체를 전달하는 것은 동작하지 않습니다
// 사실 setTimeout 코드는 this가 window 객체가 될 것으로 예상하기 때문에 오류가 발생합니다
setTimeout.call(myArray, myArray.myMethod, 2000); // error: "NS_ERROR_XPC_BAD_OP_ON_WN_PROTO: Illegal operation on WrappedNative prototype object"
setTimeout.call(myArray, myArray.myMethod, 2500, 2); // 위와 동일한 에러가 발생합니다
```

보시다시피 레거시 JavaScript에서는 `this` 객체를
콜백 함수에 전달할 방법이 없습니다.

### 가능한 해결 방법

> 모든 최신 JavaScript 런타임(브라우저 및 기타)은 lexical `this`를 사용하여 [화살표 함수](/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions)를 지원합니다. `myArray` 메서드 내부에 있는 경우 `setInterval(() => this.myMethod())`를 작성할 수 있습니다.

Reviewer: ```suggestion
모든 최신 JavaScript 런타임(브라우저 및 기타)은 lexical `this`를 사용하여 [화살표 함수](/ko/docs/Web/JavaScript/Reference/Functions/Arrow_functions)를 지원합니다. `myArray` 메서드 내부에 있는 경우 `setInterval(() => this.myMethod())`를 작성할 수 있습니다.
```

> IE를 지원해야 하는 경우 지정된 함수의 모든 호출에 `this`의 값을 고정할 수 있는 [`Function.prototype.bind()`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind) 메서드를 사용하세요. 이렇게 하면 함수가 호출된 컨텍스트에 따라 `this`가 무엇인지 명확하지 않은 문제를 쉽게 우회할 수 있습니다.

Reviewer: ```suggestion
IE를 지원해야 하는 경우 지정된 함수의 모든 호출에 `this`의 값을 고정할 수 있는 [`Function.prototype.bind()`](/ko/docs/Web/JavaScript/Reference/Global_Objects/Function/bind) 메서드를 사용하세요. 이렇게 하면 함수가 호출된 컨텍스트에 따라 `this`가 무엇인지 명확하지 않은 문제를 쉽게 우회할 수 있습니다.
```

## 사용 일람

`setInterval()` 함수는 일반적으로 애니메이션과 같이 반복해서 실행되는 함수의
지연을 설정하는 데 사용됩니다. {{domxref("clearInterval()")}}을 사용하여
interval을 취소할 수 있습니다.

지정된 지연 후 함수를 "한번" 호출하려면 {{domxref("setTimeout()")}}를
사용하세요.

### 지연 제한

간격(interval)은 중첩될 수 있습니다. 즉, `setInterval()`에 대한 콜백이
`setInterval()`을 호출하여 첫 번째 간격이 계속 진행 중일지라도
다른 간격의 실행을 시작할 수 있습니다. 이것이 성능에 미칠 수 있는 잠재적인
영향을 완화하기 위해 간격이 5개 수준 이상으로 중첩되면 브라우저는 자동으로
간격에 대해 4 ms 최소 값을 적용합니다. 중첩 호출이 심화된 `setInterval()`의
호출에서 4 ms 미만의 값을 지정하면 4 ms로 고정됩니다.

브라우저는 일부 상황에서 간격에 대해 훨씬 더 엄격한 최소 값을 적용할 수

> 있지만 일반적이지 않아야 합니다. 또한 콜백 호출 사이의 실제 경과 시간은

Reviewer: ```suggestion
있지만 이러한 경우는 일반적이지 않아야 합니다. 또한 콜백 호출 사이의 실제 경과 시간은
```
조금 더 이해하기 쉽도록 제안드려봅니다.

주어진 `delay` 보다 길 수 있습니다.

> 더 많은 예제는 [지정된 값보다 더 오래 지연되는 이유](/en-US/docs/Web/API/setTimeout#reasons_for_delays_longer_than_specified)를 참고하세요.

Reviewer: ```suggestion
더 많은 예제는 [지정된 값보다 더 오래 지연되는 이유](/ko/docs/Web/API/setTimeout#딜레이가_지정한_값보다_더_긴_이유)를 참고하세요.
```
